README.md
# SQL Abstraction Library

 * Last tests against master on CircleCI: [![CircleCI](https://circleci.com/gh/faust64/someql.svg?style=svg)](https://circleci.com/gh/faust64/someql)

 * Install with: `npm install someql`

 * Write backend-agnostic code:

```
// cassandra
const db = require'someql')({ driver: 'cassandra', database: 'tests', username: 'casuser', password: 'caspass', host: '3.4.5.6' });
// mysql
const db = require('someql')({ driver: 'mysql', database: 'tests', username: 'myuser', password: 'mypass', host: '2.3.4.5' });
// postgres
const db = require('someql')({ driver: 'postgres', database: 'tests', username: 'pguser', password: 'pgpass', host: '1.2.3.4', ssl: true });
// defaults to sqlite
const db = require('someql')({ database: './default.sqlite' });

db.read(`SELECT * from sometable`)
    .then((rsp) => {
	    if (rsp !== undefined && rsp[0] !== undefined) {
		for (let i = 0, i < rsp.length; i++) {
		    doSomethingWith(rsp[0]);
		}
		return db.write(`INSERT INTO othertable (name, otherfield) VALUES ('toto', false)`);
	    }
	})
    .then(() => whatver())
    .catch((e) => {
	    console.error('failed querying DB');
	    console.error(e);
	});
```


CHANGELOG.md
## v0.1.2 - 202009030952

 - fixing some vulnerable dependency

## v0.1.1 - 202003271218

 - fixing some dependency issue

## v0.1.0 - 202003271212

 - drops NodeJS 6 & NodeJS 8 support

## v0.0.5 - 201907131642

 - refactor allowing for multiple clients being used concurrently

## v0.0.4 - 201806110000

 - postgres connectionx cleanup

## v0.0.3 - 201801091627

 - actually test & fix cassandra implementation

## v0.0.2 - 201801091506

 - adding unit tests

## v0.0.1 - 201801010242

 - initial release


tests/main.js
describe('test database connector', () => {
	if (process.env.TEST_SQLITE) {
	    it('sqlite read & write', (done) => {
		    let db = require('../index.js')({ driver: 'sqlite' });
		    db.write('CREATE TABLE unittest (id INTEGER, value TEXT)')
			.then(() => db.write(`INSERT INTO unittest VALUES (1, 'toto')`))
			.then(() => db.read('SELECT * FROM unittest'))
			.then((rslt) => {
				if (rslt.length !== 1) {
				    throw new Error(`sqlite returned ${rslt.length} instead of our single test record`);
				} else if (rslt[0].id !== 1 || rslt[0].value !== 'toto') {
				    throw new Error(`sqlite returned inconsistent record back from our test table`);
				}
				return db.write('DROP TABLE unittest');
			    })
			.then(() => done())
			.catch((e) => done(e))
		});
	}
	if (process.env.TEST_MYSQL) {
	    it('mysql read & write', (done) => {
		    let db = require('../index.js')({ driver: 'mysql' });
		    db.write('CREATE TABLE unittest (id INTEGER, value VARCHAR(32))')
			.then(() => db.write(`INSERT INTO unittest VALUES (1, 'toto')`))
			.then(() => db.read('SELECT * FROM unittest'))
			.then((rslt) => {
				if (rslt.length !== 1) {
				    throw new Error(`mysql returned ${rslt.length} instead of our single test record`);
				} else if (rslt[0].id !== 1 || rslt[0].value !== 'toto') {
				    throw new Error(`mysql returned inconsistent record back from our test table`);
				}
				return db.write('DROP TABLE unittest');
			    })
			.then(() => done())
			.catch((e) => done(e))
		});
	}
	if (process.env.TEST_POSTGRES) {
	    it('postgres read & write', (done) => {
		    let db = require('../index.js')({ driver: 'postgres' });
		    db.write('CREATE TABLE unittest (id INT, value VARCHAR(32))')
			.then(() => db.write(`INSERT INTO unittest VALUES (1, 'toto')`))
			.then(() => db.read('SELECT * FROM unittest'))
			.then((rslt) => {
				if (rslt.length !== 1) {
				    throw new Error(`postgres returned ${rslt.length} instead of our single test record`);
				} else if (rslt[0].id !== 1 || rslt[0].value !== 'toto') {
				    throw new Error(`postgres returned inconsistent record back from our test table`);
				}
				return db.write('DROP TABLE unittest');
			    })
			.then(() => done())
			.catch((e) => done(e))
		});
	}
	if (process.env.TEST_CASSANDRA) {
	    it('cassandra read & write', (done) => {
		    let db = require('../index.js')({ driver: 'cassandra' });
		    db.write('CREATE TABLE unittest (id INT PRIMARY KEY, value VARCHAR)')
			.then(() => db.write(`INSERT INTO unittest (id, value) VALUES (1, 'toto')`))
			.then(() => db.read('SELECT * FROM unittest'))
			.then((rslt) => {
				if (rslt.length !== 1) {
				    throw new Error(`cassandra returned ${rslt.length} instead of our single test record`);
				} else if (rslt[0].id !== 1 || rslt[0].value !== 'toto') {
				    throw new Error(`cassandra returned inconsistent record back from our test table`);
				}
				return db.write('DROP TABLE unittest');
			    })
			.then(() => done())
			.catch((e) => done(e))
		});
	}
    });


mysql.js
const Promise = require('bluebird');
const logger = require('wraplog')('mysql-handler');

class MySQL {
    constructor(opts) {
	try {
	    let mysql = require('mysql'), myopts = {
		    connectionLimit: opts.maxConn || 64,
		    database: opts.database || 'default',
		    host: opts.host || '127.0.0.1',
		    insecureAuth: process.env.MYSQL_INSECURE_AUTH !== undefined,
		    port: opts.port || 3306,
		    user: opts.username || 'root'
		};
	    if (opts.password !== undefined) { myopts.password = opts.password; }
	    Promise.promisifyAll(mysql);
	    Promise.promisifyAll(require("mysql/lib/Connection").prototype);
	    Promise.promisifyAll(require("mysql/lib/Pool").prototype);
	    this._db = mysql.createPool(myopts);
	    this._db.on('connection', connection => connection.query('SET SESSION auto_increment_increment=1'));
	    this._db.on('enqueue', () => logger.info('waiting for mysql socket'));
	    this._opts = opts;
	    let self = this;
	    this._getSocket = () => { return self._db.getConnectionAsync().disposer(connection => connection.release()) };
	} catch(e) {
	    logger.error(e);
	    process.exit(1);
	}
    }

    get handlers() {
	return {
		write: (qry) => {
			let self = this;
			return Promise.using(self._getSocket(), socket => {
				    logger.debug(qry);
				    return socket.queryAsync(qry)
				})
			    .catch((e) => {
				    logger.error(e);
				    throw { code: 500, msg: 'failed writing to database' };
				});
		    },
		read: (qry, limit, offset) => {
			let self = this;
			return Promise.using(self._getSocket(), socket => {
				    let mylimit = limit || self._opts.paginationMin || 100;
				    let myoffset = offset || 0;
				    if (qry.indexOf(' LIMIT ') < 0 && mylimit !== 'none') {
					qry += ` LIMIT ${myoffset}, ${mylimit}`;
				    }
				    logger.debug(qry);
				    return socket.queryAsync(qry);
				})
			    .then((resp) => {
				    if (resp !== undefined && resp.length >= 0) {
					return resp;
				    } else {
					logger.error('malformatted response object from db');
					logger.error(resp);
					throw { code: 500, msg: 'malformatted response object from db' };
				    }
				})
			    .catch((e) => {
				    logger.error(e);
				    throw { code: 500, msg: 'failed querying database' };
				});
		    },
		close: () => {
			let self = this;
			self._db.endAsync();
			self._db = undefined;
		    }
	    };
    }
}

module.exports = (opts) => new MySQL(opts).handlers;


sqlite.js
const Promise = require('bluebird');
const logger = require('wraplog')('sqlite-handler');

class Sqlite {
    constructor(opts) {
	try {
	    const sqlite = require('better-sqlite3');
	    this._db = new sqlite(opts.database || './default.sqlite');
	    this._opts = opts;
	} catch(e) {
	    logger.error(e);
	    process.exit(1);
	}
    }

    get handlers() {
	return {
		write: (qry) => {
			let self = this;
			return new Promise((resolve, reject) => {
				try {
				    logger.debug(qry);
				    let r = self._db.exec(qry);
				    resolve(true);
				} catch(e) {
				    logger.error(e);
				    reject({ code: 500, msg: 'failed writing to database' });
				}
			    });
		    },
		read: (qry, limit, offset) => {
			let self = this;
			return new Promise((resolve, reject) => {
				try {
				    let mylimit = limit || self._opts.paginationMin || 100;
				    let myoffset = offset || 0;
				    if (qry.indexOf(' LIMIT ') < 0 && mylimit !== 'none') {
					qry += ` LIMIT ${myoffset}, ${mylimit}`;
				    }
				    logger.debug(qry);
				    let r = self._db.prepare(qry).all();
				    resolve(r);
				} catch(e) {
				    logger.error(e);
				    reject({ code: 500, msg: 'failed querying database' });
				}
			    });
		    },
		close: () => {
			let self = this;
			self._db.close();
			self._db = undefined;
		    }
	    };
    }
}

module.exports = (opts) => new Sqlite(opts).handlers;


postgres.js
const Promise = require('bluebird');
const logger = require('wraplog')('postgres-handler');

class Postgres {
    constructor(opts) {
	try {
	    const pool = require('pg-pool');
	    let popts = {
		    connectionTimeoutMillis: opts.connTimeout || 8000,
		    database: opts.database || 'template42',
		    host: opts.host || '127.0.0.1',
		    idleTimeoutMillis: opts.idleTimeout || 12000,
		    max: opts.maxConn || 256,
		    min: 32,
		    port: opts.port || 5432,
		    user: opts.username || 'postgres'
		};
	    if (opts.password !== undefined) { popts.password = opts.password; }
	    if (opts.ssl !== undefined) { popts.ssl = opts.ssl; }
	    this._db = new pool(popts);
	    this._opts = opts;
	} catch(e) {
	    logger.error(e);
	    process.exit(1);
	}
    }

    get handlers() {
	return {
		write: (qry) => {
			let self = this;
			return new Promise((resolve, reject) => {
				let cleanup = false;
				self._db.connect()
				    .then((socket) => {
					    cleanup = socket;
					    logger.debug(qry);
					    return socket.query(qry);
					})
				    .then(() => cleanup.release())
				    .then(() => resolve(true))
				    .catch((e) => {
					    if (cleanup !== false) { cleanup.release(); }
					    logger.error(e);
					    reject({ code: 500, msg: 'failed writing to database' });
					});
			    });
		    },
		read: (qry, limit, offset) => {
			let self = this;
			return new Promise((resolve, reject) => {
				let cleanup = false;
				self._db.connect()
				    .then((socket) => {
					    let mylimit = limit || self._opts.paginationMin || 100;
					    let myoffset = offset || 0;
					    if (qry.indexOf(' LIMIT ') < 0 && mylimit !== 'none') {
						qry += ` LIMIT ${mylimit} OFFSET ${myoffset}`;
					    }
					    cleanup = socket;
					    logger.info(qry);
					    return socket.query(qry);
					})
				    .then((resp) => {
					    cleanup.release();
					    if (resp.rows !== undefined && resp.rows.length >= 0) {
						resolve(resp.rows);
					    } else {
						logger.error('malformatted response object from db');
						logger.error(resp.rows);
						reject({ code: 500, msg: 'malformatted response object from db' });
					    }
					})
				    .catch((e) => {
					    if (cleanup !== false) { cleanup.release(); }
					    logger.error(e);
					    reject({ code: 500, msg: 'failed querying database' });
					});
			    });
		    },
		close: () => {
			let self = this;
			self._db.endAsync();
			self._db = undefined;
		    }
	    };
    }
}

module.exports = (opts) => new Postgres(opts).handlers;


cassandra.js
const Promise = require('bluebird');
const logger = require('wraplog')('cassandra-handler');

class Cassandra {
    constructor(opts) {
	let copts = opts || { contactPoints: [ '127.0.0.1' ], keyspace: 'sessions' };
	try {
	    this._drv = require('cassandra-driver');
	    if (opts.username !== undefined && opts.password !== undefined) {
		copts = new this._drv.auth.PlainTextAuthProvider(opts.username, opts.password);
	    }
	    this._db = new this._drv.Client(copts);
	    this._opts = opts;
	} catch(e) {
	    logger.error(e);
	    process.exit(1);
	}
	let self = this;
	this._resolveConsistency = (str) => {
		if (str === 'any' || str === 'ANY') { return self._drv.types.consistencies.any; }
		else if (str === 'one' || str === 'ONE') { return self._drv.types.consistencies.one; }
		else if (str === 'two' || str === 'TWO') { return self._drv.types.consistencies.two; }
		else if (str === 'three' || str === 'THREE') { return self._drv.types.consistencies.three; }
		else if (str === 'quorum' || str === 'QUORUM') { return self._drv.types.consistencies.quorum; }
		else if (str === 'all' || str === 'ALL') { return self._drv.types.consistencies.all; }
		else if (str === 'localQuorum' || str === 'LOCAL_QUORUM') { return self._drv.types.consistencies.localQuorum; }
		else if (str === 'eachQuorum' || str === 'EACHL_QUORUM') { return self._drv.types.consistencies.eachQuorum; }
		else if (str === 'serial' || str === 'SERIAL') { return self._drv.types.consistencies.serial; }
		else if (str === 'localSerial' || str === 'LOCAL_SERIAL') { return self._drv.types.consistencies.localSerial; }
		else if (str === 'localOne' || str === 'LOCAL_ONE') { return self._drv.types.consistencies.localOne; }
		else { return self._drv.types.consistencies.one; }
	    };
	this._fmtConsistency = (str) => {
		const policy = self._resolveConsistency(str);
		return { consistency: policy };
	    };
    }

    get handlers() {
	return {
		write: (qry) => {
			let self = this;
			return new Promise((resolve, reject) => {
				logger.debug(qry);
				self._db.execute(qry, [], self._fmtConsistency(self._opts.writeConsistency))
				    .then((resp) => { resolve(true); })
				    .catch((e) => {
					    logger.error(e);
					    reject('failed writing to database');
					});
			    });
		    },
	       read: (qry, limit, offset) => {
			let self = this;
			return new Promise((resolve, reject) => {
				let mylimit = limit || self._opts.paginationMin || 100;
				if (qry.indexOf(' LIMIT ') < 0 && mylimit !== 'none') {
				    qry += " LIMIT " + mylimit;
				}
				logger.debug(qry);
				self._db.execute(qry, [], self._fmtConsistency(self._opts.readConsistency))
				    .then((resp) => {
					    if (resp.rows !== undefined && resp.rows.length >= 0) {
						resolve(resp.rows);
					    } else {
						logger.error('malformatted response object from db');
						logger.error(resp.rows);
						reject('malformatted response object from db');
					    }
					})
				    .catch((e) => {
					    logger.error(e);
					    reject('failed writing to database');
					});
			});
		    },
		close: () => {
			let self = this;
			self._db.closeAsync();
			self._db = undefined;
		    }
	    };
    }
}

module.exports = (opts) => new Cassandra(opts).handlers;


index.js
const valid = [ 'cassandra', 'mysql', 'postgres', 'sqlite' ];

module.exports = (opts) => {
	let driver = 'sqlite';
	opts = opts || {};
	if (opts.driver !== undefined) {
	    if (valid.indexOf(opts.driver) >= 0) { driver = opts.driver; }
	} else if (process.env.DB_CONNECTOR !== undefined) {
	    if (valid.indexOf(process.env.DB_CONNECTOR) >= 0) { driver = process.env.DB_CONNECTOR; }
	}
	if (opts.paginationMin === undefined) { opts.paginationMin = process.env.PAGINATION_MIN || 100; }
	if (driver === 'sqlite') {
	    if (opts.database === undefined) { opts.database = process.env.SQLITE_DBFILE || './default.sqlite'; }
	    return require('./sqlite')(opts);
	} else if (driver === 'cassandra') {
	    if (opts.host === undefined) { opts.host = process.env.CASSANDRA_HOST || [ '127.0.0.1' ]; }
	    if (typeof opts.host === 'string') { opts.contactPoints = opts.host.split(','); }
	    else { opts.contactPoints = opts.host; }
	    if (opts.database === undefined) { opts.keyspace = process.env.CASSANDRA_KEYSPACE || 'sessions'; }
	    else if (opts.keyspace === undefined) { opts.keyspace = opts.database; }
	    if (opts.readConsistency === undefined) { opts.readConsistency = process.env.CASSANDRA_READ_CONSISTENCY || 'one'; }
	    if (opts.writeConsistency === undefined) { opts.writeConsistency = process.env.CASSANDRA_WRITE_CONSISTENCY || 'one'; }
	    if (opts.username === undefined && process.env.CASSANDRA_USER !== undefined) { opts.username = process.env.CASSANDRA_USER; }
	    if (opts.password === undefined && process.env.CASSANDRA_PASSWORD !== undefined) { opts.password = process.env.CASSANDRA_PASSWORD; }
	    return require('./cassandra.js')(opts);
	} else if (driver === 'mysql') {
	    if (opts.database === undefined) { opts.database = process.env.MYSQL_DATABASE || 'sessions'; }
	    if (opts.maxConn === undefined) { opts.maxConn = process.env.MYSQL_MAX_CONN || 64; }
	    if (opts.host === undefined) { opts.host = process.env.MYSQL_HOST || '127.0.0.1'; }
	    if (opts.password === undefined && process.env.MYSQL_PASSWORD !== undefined) { opts.password = process.env.MYSQL_PASSWORD; }
	    if (opts.port === undefined) { opts.port = process.env.MYSQL_PORT || 3306; }
	    if (opts.username === undefined) { opts.username = process.env.MYSQL_USER || 'root'; }
	    return require('./mysql.js')(opts);
	} else if (driver === 'postgres') {
	    if (opts.connTimeout === undefined) { opts.connTimeout = process.env.POSTGRES_CONN_TIMEOUT || 1000; }
	    if (opts.database === undefined) { opts.database = process.env.POSTGRES_DATABASE || 'sessions'; }
	    if (opts.maxConn === undefined) { opts.maxConn = process.env.POSTGRES_MAX_CONN || 64; }
	    if (opts.idleTimeout === undefined) { opts.idleTimeout = process.env.POSTGRES_IDLE_TIMEOUT || 1000; }
	    if (opts.host === undefined) { opts.host = process.env.POSTGRES_HOST || '127.0.0.1'; }
	    if (opts.password === undefined && process.env.POSTGRES_PASSWORD !== undefined) { opts.password = process.env.POSTGRES_PASSWORD; }
	    if (opts.port === undefined) { opts.port = process.env.POSTGRES_PORT || 5432; }
	    if (opts.username === undefined) { opts.username = process.env.POSTGRES_USER || 'postgres'; }
	    if (opts.ssl === undefined) { opts.ssl = (process.env.POSTGRES_SSL !== undefined && process.env.POSTGRES_SSL !== 'false'); }
	    return require('./postgres.js')(opts);
	}
	return new Error('unsupported DB driver ' + driver);
    };
